release-23.2: sql: block concurrent renames with declarative schema changes

fqazi opened the pull request:
Backport 1/1 commits from #128683.

/cc @cockroachdb/release

---

Previously, we allowed the legacy and declarative schema change jobs to run concurrently with declarative schema change jobs, for example renames of objects. If a drop operation was executed next to a legacy schema change job, it was possible for the legacy code to delete the descriptor before the declarative schema changer. To address this, we are going to safe guard rename operations, which currently do not check for declarative schema changer jobs.

Fixes: #128268

Release note (bug fix): Declarative and legacy schema changes were
incorrectly allowed to be executed concurrently before, which could lead
to failing or hung schema change jobs.

Release justification: bug fix

fqazi committed:
sql: block concurrent renames with declarative schema changes

Previously, we allowed the legacy and declarative schema change jobs to
run concurrently with declarative schema change jobs, for example
renames of objects. If a drop operation was executed next to a legacy
schema change job, it was possible for the legacy code to delete the
descriptor before the declarative schema changer. To address this, we
are going to safe guard rename operations, which currently do not check
for declarative schema changer jobs.

Fixes: #128268

Release note (bug fix): Declarative and legacy schema changes were
incorrectly allowed to be executed concurrently before, which could lead
to failing or hung schema change jobs.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/128838)
<!-- Reviewable:end -->

fqazi committed:
sql/schemachanger: fix flake inside TestConcurrentSchemaChanges

Previously, this test would flake if descriptors were being dropped
concurrently while dropping tables. This was because #128683 added
support for dropping schemas. To address this, this patch adds the
descriptor being dropped as a valid error for CREATE INDEX.
Additionally, this patch fixes the rename logic to avoid collisions by
sequentially picking object names.

Fixes: #128731, #128797

Release note: None
